router/pkg/pubsub/kafka/kafka.go

```diff
@@ -0,0 +1,380 @@
+		return nil, fmt.Errorf("failed to create write client for Kafka: %w", err)
+	}
+
+	return &connector{
+		writeClient: writeClient,
+		opts:        opts,
+		logger:      logger,
+	}, nil
+}
+
+func (c *connector) New(ctx context.Context) pubsub_datasource.KafkaPubSub {
+	ps := &kafkaPubSub{
+		ctx:           ctx,
+		logger:        c.logger.With(zap.String("pubsub", "kafka")),
+		opts:          c.opts,
+		subscriptions: make(map[string][]*subscription),
+		clients:       make(map[string]*kgo.Client),
+		writeClient:   c.writeClient,
+		mu:            sync.RWMutex{},
+		closeWg:       sync.WaitGroup{},
+	}
+
+	return ps
+}
+
+// kafkaPubSub is a Kafka pubsub implementation.
+// It uses the franz-go Kafka client to consume and produce messages.
+// The pubsub is stateless and does not store any messages.
+// It uses a single write client to produce messages and a client per topic to consume messages.
+// Each client polls the Kafka topic for new records and updates the subscriptions with the new data.
+type kafkaPubSub struct {
+	ctx           context.Context
+	opts          []kgo.Opt
+	logger        *zap.Logger
+	clients       map[string]*kgo.Client
+	subscriptions map[string][]*subscription
+	writeClient   *kgo.Client
+	mu            sync.RWMutex
+	closeWg       sync.WaitGroup
+}
+
+// addSubscription adds the subscription to the subscriptions map.
+func (p *kafkaPubSub) addSubscription(sub *subscription) {
+
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	for _, topic := range sub.topics {
+		if _, ok := p.subscriptions[topic]; !ok {
+			p.subscriptions[topic] = make([]*subscription, 0, 10)
+		}
+		p.subscriptions[topic] = append(p.subscriptions[topic], sub)
+	}
+}
+
+// topicPoller polls the Kafka topic for new records and calls the updateTriggers function.
+func (p *kafkaPubSub) topicPoller(client *kgo.Client) error {
+	p.closeWg.Add(1)
+	defer p.closeWg.Done()
+
+	for {
+		select {
+		// Close the client if the context was canceled
+		// This is not the context of the subscription, but the context of application shutdown
+		case <-p.ctx.Done():
+			client.Close()
+			return nil
+
+		default:
+			// Try to fetch max records from any subscribed topics
+			// In the future, we could create a client per topic to fetch in parallel
+			fetches := client.PollRecords(p.ctx, 10_000)
+			if fetches.IsClientClosed() {
+				return errClientClosed
+			}
+
+			if errs := fetches.Errors(); len(errs) > 0 {
+
+				for _, fetchError := range errs {
+
+					// If the context was canceled, the error is wrapped in a fetch error
+					if errors.Is(fetchError.Err, context.Canceled) {
+						return nil
+					}
+
+					var kErr *kerr.Error
+					if errors.As(fetchError.Err, &kErr) {
+						if !kErr.Retriable {
+							p.logger.Error("unrecoverable fetch error",
+								zap.Error(fetchError.Err),
+								zap.String("topic", fetchError.Topic),
+							)
+
+							// If the error is not recoverable, return it and abort the poller
+							return fetchError.Err
```

Comment:
Are you sure that we want to kill all Subscriptions that depend on this poller when such an error occurs? What errors are possible? Should we handle this more gracefully?

Reply:
Yes, the library retries in case of retriable errors. If an unretriable errors occurs e.g. misconfigurations or misbehavior of the cluster side can't be replied. If we would swallow it, it would produce a lot logs.